### PR TITLE
Audio Stem data to shaders + "Twist" axis swap button.

### PR DESCRIPTION
**Pull request by zranger1**
Shader patterns and effects can now access audio stem data via the following uniforms:
```
uniform float stemBass;
uniform float stemDrums;
uniform float stemVocal;
uniform float stemOther;
```
Also added "**Twist**" feature (x/z axis swapping) as discussed.  It currently works for all shader patterns and is controlled by a design-mode button.

### CODE REVIEW

**Review by jkbelcher — COMMENTED** [review]
High-level looks good.   Added a few comments about details.

File: src/main/java/titanicsend/pattern/TECommonControls.java [hunk 1]
```diff
@@ -122,6 +122,12 @@ public void buildDefaultControlList() {
             .setDescription("Trigger WoW effects");
     setControl(TEControlTag.WOWTRIGGER, p);
 
+    p =
+      new BooleanParameter(TEControlTag.TWIST.getLabel(), false)
+        .setMode(BooleanParameter.Mode.MOMENTARY)
```
[the remark this answers is not in the document]
Comment (worldveil) [reply src/main/java/titanicsend/pattern/TECommonControls.java:127]: +1

[the remark this answers is not in the document]
Comment (zranger1) [reply src/main/java/titanicsend/pattern/TECommonControls.java]: Excellent thought.  Done!

File: src/main/java/titanicsend/pattern/glengine/GLEffectControl.java [hunk 2]
```diff
@@ -27,6 +28,12 @@ public void setUniforms(GLShader s) {
     s.setUniform("iTime", (float) effect.getTime());
     s.setUniform("iResolution", (float) GLEngine.getWidth(), (float) GLEngine.getHeight());
 
+    // Current values from audio stems
+    s.setUniform("stemBass", (float) AudioStems.current.bass.getValue());
```
[the remark this answers is not in the document]
Comment (zranger1) [reply src/main/java/titanicsend/pattern/glengine/GLEffectControl.java]: You're right  - same effect, less typing!   Done.

File: src/main/java/titanicsend/pattern/glengine/GLPatternControl.java [hunk 3]
```diff
@@ -31,6 +32,12 @@ public void setUniforms(GLShader s) {
     s.setUniform("levelReact", (float) pattern.getLevelReactivity());
     s.setUniform("frequencyReact", (float) pattern.getFrequencyReactivity());
 
+    // Current values from audio stems
+    s.setUniform("stemBass", (float) AudioStems.current.bass.getValue());
```
[the remark this answers is not in the document]
Comment (zranger1) [reply src/main/java/titanicsend/pattern/glengine/GLPatternControl.java]: Done

File: src/main/java/titanicsend/pattern/glengine/ShaderPaint2d.java [hunk 4]
```diff
@@ -22,10 +22,12 @@ public ShaderPaint2d(boolean isStaticModel) {
...
   public void mapToPoints(List<LXPoint> points, ByteBuffer image, int[] colors) {
-    // TODO - remove this block when we move completely to dynamic model
-    // TODO - (as well as the 'Static' versions of these functions)
-    if (isStatic) {
-      mapToPointsStatic(points, image, colors);
+    // Added "twist" feature - allows controlled swap of x and z axes.  This must
+    // work on both static and dynamic models for now, which makes the logic a little
+    // odd to look at. So...
+    // TODO - simplify twist logic when we finish move to dynamic model
+    if ((isTwisted && !isStatic) || isStatic && !isTwisted) {
```
[the remark this answers is not in the document]
Comment (zranger1) [reply src/main/java/titanicsend/pattern/glengine/ShaderPaint2d.java]: Ok, you made me think about how this logic actually works:
```
Twisted   Static   Map
--------------------------
true       true     normal
false      true     ZX
true       false    ZX
false      false    normal 
```
In code reduces to (isTwisted != isStatic), which is much simpler!

File: src/main/java/titanicsend/pattern/glengine/ShaderPaint2d.java [hunk 5]
```diff
@@ -52,10 +54,9 @@ public void mapToPoints(List<LXPoint> points, ByteBuffer image, int[] colors) {
    * @param colors array of colors, one for each point
    */
   public void mapToBuffer(List<LXPoint> points, ByteBuffer image, int[] colors) {
-    // TODO - remove this block when we move completely to dynamic model
-    // TODO - (as well as the 'Static' versions of these functions)
-    if (isStatic) {
-      mapToBufferStatic(points, image, colors);
+    // TODO - simplify twist logic when we finish move to dynamic model
+    if ((isTwisted && !isStatic) || isStatic && !isTwisted) {
```
[the remark this answers is not in the document]
Comment (zranger1) [reply src/main/java/titanicsend/pattern/glengine/ShaderPaint2d.java]: Fixed in latest commit

File: src/main/java/titanicsend/pattern/glengine/ShaderPaint3d.java [hunk 6]
```diff
@@ -24,11 +24,12 @@ public ShaderPaint3d(boolean isStaticModel) {
...
-
-    // TODO - remove this block when we move completely to dynamic model
-    // TODO - (as well as the 'Static' versions of these functions)
-    if (isStatic) {
-      mapToPointsStatic(points, image, colors);
+    // Added "twist" feature - allows controlled swap of x and z axes.  This must
+    // work on both static and dynamic models for now, which makes the logic a little
+    // odd to look at. So...
+    // TODO - simplify twist logic when we finish move to dynamic model
+    if ((isTwisted && !isStatic) || isStatic && !isTwisted) {
```
[the remark this answers is not in the document]
Comment (zranger1) [reply src/main/java/titanicsend/pattern/glengine/ShaderPaint3d.java]: Fixed in latest commit

File: src/main/java/titanicsend/pattern/glengine/ShaderPaint3d.java [hunk 7]
```diff
@@ -48,11 +49,9 @@ public void mapToPoints(List<LXPoint> points, ByteBuffer image, int[] colors) {
   }
 
   public void mapToBuffer(List<LXPoint> points, ByteBuffer image, int[] colors) {
-
-    // TODO - remove this block when we move completely to dynamic model
-    // TODO - (as well as the 'Static' versions of these functions)
-    if (isStatic) {
-      mapToBufferStatic(points, image, colors);
+    // TODO - simplify twist logic when we finish move to dynamic model
+    if ((isTwisted && !isStatic) || isStatic && !isTwisted) {
```
[the remark this answers is not in the document]
Comment (zranger1) [reply src/main/java/titanicsend/pattern/glengine/ShaderPaint3d.java]: Fixed in latest commit

**Review by andrewlook — APPROVED** [review]
awesome! can't wait to try writing a shader to use the raw stems :)

File: src/main/java/titanicsend/pattern/glengine/ShaderPaint2d.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (andrewlook) [reply src/main/java/titanicsend/pattern/glengine/ShaderPaint2d.java]: thanks for spelling out the logic here 👍

**Pull request merged**